File: Zeni/android/app/src/main/java/com/zeni/voiceai/protocol/Protocol.kt
package com.zeni.voiceai.protocol

import com.google.gson.annotations.SerializedName
import java.util.UUID

/**
 * WebSocket message types for Zeni protocol.
 */
object MessageType {
    const val SESSION_START = "session_start"
    const val SESSION_END = "session_end"
    const val SESSION_ACK = "session_ack"
    const val LANGUAGE_CHANGE = "language_change"
    const val VOICE_CHANGE = "voice_change"
    const val TTS_PROVIDER_CHANGE = "tts_provider_change"
    const val TTS_SPEED_CHANGE = "tts_speed_change"
    const val PERSONALITY_CHANGE = "personality_change"
    const val AUDIO_FRAME = "audio_frame"
    const val IMAGE_FRAME = "image_frame"  // Camera frame for visual context
    const val REQUEST_IMAGE = "request_image"  // Server requests image capture
    const val SPEECH_FINISHED = "speech_finished"
    const val AUDIO_RESPONSE = "audio_response"
    const val TRANSCRIPT_PARTIAL = "transcript_partial"
    const val TRANSCRIPT_FINAL = "transcript_final"
    const val LLM_TOKEN = "llm_token"
    const val LLM_COMPLETE = "llm_complete"
    const val INTERRUPT = "interrupt"
    const val PLAYBACK_STOP = "playback_stop"
    const val ERROR = "error"
    const val HEARTBEAT = "heartbeat"
    const val HEARTBEAT_ACK = "heartbeat_ack"
    const val PING = "ping"
    const val PONG = "pong"
    const val STATE_CHANGE = "state_change"
    const val CAMPUS_TOUR = "campus_tour"  // For virtual campus tours
    const val FEE_STRUCTURE = "fee_structure"  // For detailed fee structure display
    const val SHOW_PLACEMENTS = "show_placements"  // For top placement students gallery
    const val ROBOT_COMMAND = "robot_command"  // Robot movement commands from LLM
    const val ROBOT_STATUS = "robot_status"  // Client reports robot connection status
}

/**
 * Session states.
 */
enum class SessionState {
    @SerializedName("idle") IDLE,
    @SerializedName("listening") LISTENING,
    @SerializedName("transcribing") TRANSCRIBING,
    @SerializedName("generating") GENERATING,
    @SerializedName("speaking") SPEAKING,
    @SerializedName("interrupted") INTERRUPTED,
    @SerializedName("error") ERROR,
    @SerializedName("closed") CLOSED
}

/**
 * Supported languages.
 */
enum class Language(val code: String) {
    @SerializedName("en") ENGLISH("en"),
    @SerializedName("hi") HINDI("hi"),
    @SerializedName("auto") AUTO("auto")
}

// ============== Outbound Messages (Client -> Server) ==============

/**
 * Session configuration.
 */
data class SessionConfig(
    @SerializedName("sample_rate") val sampleRate: Int = 16000,
    @SerializedName("language_preference") val languagePreference: String = "auto",
    @SerializedName("voice_preference") val voicePreference: String = "Kore",  // Female voice
    @SerializedName("tts_provider") val ttsProvider: String = "google",
    @SerializedName("speaking_rate") val speakingRate: Float = 1.0f,
    @SerializedName("push_to_talk") val pushToTalk: Boolean = false,
    @SerializedName("personality") val personality: String = "assistant"  // assistant or human
)

/**
 * Session start message.
 */
data class SessionStartMessage(
    @SerializedName("type") val type: String = MessageType.SESSION_START,
    @SerializedName("session_id") val sessionId: String = UUID.randomUUID().toString(),
    @SerializedName("config") val config: SessionConfig = SessionConfig(),
    @SerializedName("timestamp") val timestamp: Long = System.currentTimeMillis()
)

/**
 * Audio frame message.
 */
data class AudioFrameMessage(
    @SerializedName("type") val type: String = MessageType.AUDIO_FRAME,
    @SerializedName("timestamp") val timestamp: Long,
    @SerializedName("data") val data: String, // Base64 encoded PCM
    @SerializedName("sequence") val sequence: Int
)

/**
 * Interrupt message.
 */
data class InterruptMessage(
    @SerializedName("type") val type: String = MessageType.INTERRUPT,
    @SerializedName("timestamp") val timestamp: Long = System.currentTimeMillis()
)

/**
 * Speech finished message (explicit end of speech).
 */
data class SpeechFinishedMessage(
    @SerializedName("type") val type: String = MessageType.SPEECH_FINISHED,
    @SerializedName("timestamp") val timestamp: Long = System.currentTimeMillis()
)

/**
 * Heartbeat message.
 */
data class HeartbeatMessage(
    @SerializedName("type") val type: String = MessageType.HEARTBEAT,
    @SerializedName("timestamp") val timestamp: Long = System.currentTimeMillis()
)

/**
 * Language change message.
 */
data class LanguageChangeMessage(
    @SerializedName("type") val type: String = MessageType.LANGUAGE_CHANGE,
    @SerializedName("language") val language: String,
    @SerializedName("timestamp") val timestamp: Long = System.currentTimeMillis()
)

/**
 * Voice change message.
 */
data class VoiceChangeMessage(
    @SerializedName("type") val type: String = MessageType.VOICE_CHANGE,
    @SerializedName("voice") val voice: String,
    @SerializedName("timestamp") val timestamp: Long = System.currentTimeMillis()
)

/**
 * TTS Provider change message.
 */
data class TtsProviderChangeMessage(
    @SerializedName("type") val type: String = MessageType.TTS_PROVIDER_CHANGE,
    @SerializedName("provider") val provider: String,
    @SerializedName("timestamp") val timestamp: Long = System.currentTimeMillis()
)

/**
 * TTS Speed change message.
 */
data class TtsSpeedChangeMessage(
    @SerializedName("type") val type: String = MessageType.TTS_SPEED_CHANGE,
    @SerializedName("speed") val speed: Float,
    @SerializedName("timestamp") val timestamp: Long = System.currentTimeMillis()
)

/**
 * Personality change message.
 */
data class PersonalityChangeMessage(
    @SerializedName("type") val type: String = MessageType.PERSONALITY_CHANGE,
    @SerializedName("personality") val personality: String,  // "assistant" or "human"
    @SerializedName("timestamp") val timestamp: Long = System.currentTimeMillis()
)

/**
 * Session end message.
 */
data class SessionEndMessage(
    @SerializedName("type") val type: String = MessageType.SESSION_END,
    @SerializedName("session_id") val sessionId: String,
    @SerializedName("timestamp") val timestamp: Long = System.currentTimeMillis()
)

// ============== Inbound Messages (Server -> Client) ==============

/**
 * Generic message for initial parsing.
 */
data class GenericMessage(
    @SerializedName("type") val type: String
)

/**
 * Session acknowledgment.
 */
data class SessionAckMessage(
    @SerializedName("type") val type: String,
    @SerializedName("session_id") val sessionId: String,
    @SerializedName("status") val status: String,
    @SerializedName("timestamp") val timestamp: Long
)

/**
 * Audio response chunk.
 */
data class AudioResponseMessage(
    @SerializedName("type") val type: String,
    @SerializedName("sequence") val sequence: Int,
    @SerializedName("data") val data: String, // Base64 encoded PCM
    @SerializedName("final") val isFinal: Boolean,
    @SerializedName("sample_rate") val sampleRate: Int
)

/**
 * Partial transcript.
 */
data class TranscriptPartialMessage(
    @SerializedName("type") val type: String,
    @SerializedName("text") val text: String,
    @SerializedName("language") val language: String,
    @SerializedName("timestamp") val timestamp: Long
)

/**
 * Final transcript.
 */
data class TranscriptFinalMessage(
    @SerializedName("type") val type: String,
    @SerializedName("text") val text: String,
    @SerializedName("confidence") val confidence: Float,
    @SerializedName("language") val language: String,
    @SerializedName("timestamp") val timestamp: Long
)

/**
 * LLM token.
 */
data class LLMTokenMessage(
    @SerializedName("type") val type: String,
    @SerializedName("token") val token: String,
    @SerializedName("sequence") val sequence: Int
)

/**
 * LLM complete response.
 */
data class LLMCompleteMessage(
    @SerializedName("type") val type: String,
    @SerializedName("full_text") val fullText: String,
    @SerializedName("timestamp") val timestamp: Long
)

/**
 * State change notification.
 */
data class StateChangeMessage(
    @SerializedName("type") val type: String,
    @SerializedName("state") val state: String,
    @SerializedName("previous_state") val previousState: String?,
    @SerializedName("timestamp") val timestamp: Long
)

/**
 * Playback stop signal.
 */
data class PlaybackStopMessage(
    @SerializedName("type") val type: String,
    @SerializedName("timestamp") val timestamp: Long
)

/**
 * Error message.
 */
data class ErrorMessage(
    @SerializedName("type") val type: String,
    @SerializedName("code") val code: Int,
    @SerializedName("message") val message: String,
    @SerializedName("details") val details: Map<String, Any>?,
    @SerializedName("timestamp") val timestamp: Long
)

/**
 * Heartbeat acknowledgment.
 */
data class HeartbeatAckMessage(
    @SerializedName("type") val type: String,
    @SerializedName("timestamp") val timestamp: Long
)

/**
 * Campus tour message - triggers full-screen Matterport view.
 */
data class CampusTourMessage(
    @SerializedName("type") val type: String,
    @SerializedName("tour_id") val tourId: String,
    @SerializedName("name") val name: String,
    @SerializedName("url") val url: String,
    @SerializedName("description") val description: String,
    @SerializedName("timestamp") val timestamp: Long
)

/**
 * Fee structure message - triggers full-screen WebView with fee details.
 */
data class FeeStructureMessage(
    @SerializedName("type") val type: String,
    @SerializedName("program_id") val programId: String,
    @SerializedName("program_name") val programName: String,
    @SerializedName("url") val url: String,
    @SerializedName("timestamp") val timestamp: Long
)

/**
 * Placement gallery message - triggers full-screen placement photo viewer.
 */
data class PlacementMessage(
    @SerializedName("type") val type: String,
    @SerializedName("title") val title: String,
    @SerializedName("timestamp") val timestamp: Long
)

/**
 * Robot command message - movement commands from LLM for robot control.
 * 
 * Supported actions:
 * - forward: Move forward (1 sec for safety)
 * - backward: Move backward (1 sec for safety)
 * - left: Turn left
 * - right: Turn right
 * - spin_left: Spin in place to the left
 * - spin_right: Spin in place to the right
 * - stop: Stop all movement
 * - lights_on/lights_off: Toggle lights
 * - horn: Sound horn
 */
data class RobotCommandMessage(
    @SerializedName("type") val type: String,
    @SerializedName("action") val action: String,  // forward, backward, left, right, spin_left, spin_right, stop
    @SerializedName("duration_ms") val durationMs: Long = 1000,  // Duration in ms (capped at 1000 for safety)
    @SerializedName("speed_percent") val speedPercent: Int = 50,  // Speed 0-100 (default 50 for safety)
    @SerializedName("timestamp") val timestamp: Long
)

File: Zeni/server/core/protocol.py
"""
Zeni Protocol Definitions
Defines all message types and data structures for WebSocket communication.
"""

from enum import Enum
from typing import Optional, Dict, Any, List
from datetime import datetime
from pydantic import BaseModel, Field
import uuid


class MessageType(str, Enum):
    """WebSocket message types."""
    # Session control
    SESSION_START = "session_start"
    SESSION_END = "session_end"
    SESSION_ACK = "session_ack"
    LANGUAGE_CHANGE = "language_change"
    VOICE_CHANGE = "voice_change"
    TTS_PROVIDER_CHANGE = "tts_provider_change"
    TTS_SPEED_CHANGE = "tts_speed_change"
    PERSONALITY_CHANGE = "personality_change"  # Switch between assistant/human personality
    
    # Keepalive
    PING = "ping"
    PONG = "pong"

    # Audio flow
    AUDIO_FRAME = "audio_frame"
    AUDIO_RESPONSE = "audio_response"
    SPEECH_FINISHED = "speech_finished"
    
    # Vision
    IMAGE_FRAME = "image_frame"  # Camera frame for visual context
    REQUEST_IMAGE = "request_image"  # Server requests image capture from client
    
    # Transcript events
    TRANSCRIPT_PARTIAL = "transcript_partial"
    TRANSCRIPT_FINAL = "transcript_final"
    
    # LLM events
    LLM_TOKEN = "llm_token"
    LLM_COMPLETE = "llm_complete"
    
    # Control signals
    INTERRUPT = "interrupt"
    PLAYBACK_STOP = "playback_stop"
    
    # Interactive features
    CAMPUS_TOUR = "campus_tour"  # Trigger 360° campus tour
    FEE_STRUCTURE = "fee_structure"  # Show detailed fee structure
    SHOW_PLACEMENTS = "show_placements"  # Show top placement students
    
    # Robot control
    ROBOT_COMMAND = "robot_command"  # Send movement command to robot
    ROBOT_STATUS = "robot_status"  # Client reports robot connection status
    
    # Status
    ERROR = "error"
    HEARTBEAT = "heartbeat"
    HEARTBEAT_ACK = "heartbeat_ack"
    STATE_CHANGE = "state_change"


class SessionState(str, Enum):
    """Session state machine states."""
    IDLE = "idle"
    LISTENING = "listening"
    TRANSCRIBING = "transcribing"
    GENERATING = "generating"
    SPEAKING = "speaking"
    INTERRUPTED = "interrupted"
    ERROR = "error"
    CLOSED = "closed"


class Language(str, Enum):
    """Supported languages."""
    ENGLISH = "en"
    HINDI = "hi"
    AUTO = "auto"


class TTSProvider(str, Enum):
    """TTS Service Provider - Google Cloud TTS only."""
    GOOGLE = "google"


class Personality(str, Enum):
    """AI Personality modes."""
    ASSISTANT = "assistant"  # Professional college bot with RAG
    HUMAN = "human"  # Friendly college bot with RAG
    GENERAL = "general"  # Free AI chat - no RAG, no college data, just talk


# ============== Inbound Messages (Client -> Server) ==============

class SessionConfig(BaseModel):
    """Session configuration from client."""
    sample_rate: int = 16000
    language_preference: Language = Language.AUTO
    voice_preference: str = "Kore"  # Female voice
    tts_provider: TTSProvider = TTSProvider.GOOGLE
    speaking_rate: float = 1.0
    push_to_talk: bool = False
    personality: Personality = Personality.ASSISTANT  # AI personality mode


class SessionStartMessage(BaseModel):
    """Session start message."""
    type: MessageType = MessageType.SESSION_START
    session_id: str = Field(default_factory=lambda: str(uuid.uuid4()))
    config: SessionConfig = Field(default_factory=SessionConfig)
    timestamp: int = Field(default_factory=lambda: int(datetime.now().timestamp() * 1000))


class AudioFrameMessage(BaseModel):
    """Audio frame from client."""
    type: MessageType = MessageType.AUDIO_FRAME
    timestamp: int
    data: str  # Base64 encoded PCM data
    sequence: int


class InterruptMessage(BaseModel):
    """Interrupt signal from client."""
    type: MessageType = MessageType.INTERRUPT
    timestamp: int = Field(default_factory=lambda: int(datetime.now().timestamp() * 1000))


class HeartbeatMessage(BaseModel):
    """Heartbeat message."""
    type: MessageType = MessageType.HEARTBEAT
    timestamp: int = Field(default_factory=lambda: int(datetime.now().timestamp() * 1000))


class LanguageChangeMessage(BaseModel):
    """Language change message."""
    type: MessageType = MessageType.LANGUAGE_CHANGE
    language: Language
    timestamp: int = Field(default_factory=lambda: int(datetime.now().timestamp() * 1000))


class VoiceChangeMessage(BaseModel):
    type: MessageType = MessageType.VOICE_CHANGE
    voice: str
    timestamp: int = Field(default_factory=lambda: int(datetime.now().timestamp() * 1000))


class TtsProviderChangeMessage(BaseModel):
    type: MessageType = MessageType.TTS_PROVIDER_CHANGE
    provider: TTSProvider
    timestamp: int = Field(default_factory=lambda: int(datetime.now().timestamp() * 1000))


class TtsSpeedChangeMessage(BaseModel):
    type: MessageType = MessageType.TTS_SPEED_CHANGE
    speed: float
    timestamp: int = Field(default_factory=lambda: int(datetime.now().timestamp() * 1000))


class PersonalityChangeMessage(BaseModel):
    """Personality mode change message."""
    type: MessageType = MessageType.PERSONALITY_CHANGE
    personality: Personality
    timestamp: int = Field(default_factory=lambda: int(datetime.now().timestamp() * 1000))


class SessionEndMessage(BaseModel):
    """Session end message."""
    type: MessageType = MessageType.SESSION_END
    session_id: str
    timestamp: int = Field(default_factory=lambda: int(datetime.now().timestamp() * 1000))


class SpeechFinishedMessage(BaseModel):
    """Signal from client that speech has ended (e.g. PTT button released)."""
    type: MessageType = MessageType.SPEECH_FINISHED
    timestamp: int = Field(default_factory=lambda: int(datetime.now().timestamp() * 1000))


# ============== Outbound Messages (Server -> Client) ==============

class SessionAckMessage(BaseModel):
    """Session acknowledgment."""
    type: MessageType = MessageType.SESSION_ACK
    session_id: str
    status: str = "connected"
    timestamp: int = Field(default_factory=lambda: int(datetime.now().timestamp() * 1000))


class AudioResponseMessage(BaseModel):
    """Audio response chunk to client."""
    type: MessageType = MessageType.AUDIO_RESPONSE
    sequence: int
    data: str  # Base64 encoded PCM data
    final: bool = False
    sample_rate: int = 24000


class TranscriptPartialMessage(BaseModel):
    """Partial transcript update."""
    type: MessageType = MessageType.TRANSCRIPT_PARTIAL
    text: str
    language: Language
    timestamp: int = Field(default_factory=lambda: int(datetime.now().timestamp() * 1000))


class TranscriptFinalMessage(BaseModel):
    """Final transcript."""
    type: MessageType = MessageType.TRANSCRIPT_FINAL
    text: str
    confidence: float
    language: Language
    timestamp: int = Field(default_factory=lambda: int(datetime.now().timestamp() * 1000))


class LLMTokenMessage(BaseModel):
    """LLM token stream."""
    type: MessageType = MessageType.LLM_TOKEN
    token: str
    sequence: int


class LLMCompleteMessage(BaseModel):
    """LLM complete response."""
    type: MessageType = MessageType.LLM_COMPLETE
    full_text: str
    timestamp: int = Field(default_factory=lambda: int(datetime.now().timestamp() * 1000))


class PlaybackStopMessage(BaseModel):
    """Stop playback signal."""
    type: MessageType = MessageType.PLAYBACK_STOP
    timestamp: int = Field(default_factory=lambda: int(datetime.now().timestamp() * 1000))


class StateChangeMessage(BaseModel):
    """State change notification."""
    type: MessageType = MessageType.STATE_CHANGE
    state: SessionState
    previous_state: Optional[SessionState] = None
    timestamp: int = Field(default_factory=lambda: int(datetime.now().timestamp() * 1000))


class ErrorMessage(BaseModel):
    """Error message."""
    type: MessageType = MessageType.ERROR
    code: int
    message: str
    details: Optional[Dict[str, Any]] = None
    timestamp: int = Field(default_factory=lambda: int(datetime.now().timestamp() * 1000))


class HeartbeatAckMessage(BaseModel):
    """Heartbeat acknowledgment."""
    type: MessageType = MessageType.HEARTBEAT_ACK
    timestamp: int = Field(default_factory=lambda: int(datetime.now().timestamp() * 1000))


class CampusTourMessage(BaseModel):
    """Campus tour trigger - opens Matterport 360° view in Android app."""
    type: MessageType = MessageType.CAMPUS_TOUR
    tour_id: str  # Unique tour identifier
    name: str  # Human-readable name (e.g., "Seminar Hall")
    url: str  # Matterport URL
    description: str  # Brief description
    timestamp: int = Field(default_factory=lambda: int(datetime.now().timestamp() * 1000))


class FeeStructureMessage(BaseModel):
    """Fee structure trigger - opens WebView with fee details in Android app."""
    type: MessageType = MessageType.FEE_STRUCTURE
    program_id: str  # Unique program identifier (e.g., "btech-cse")
    program_name: str  # Human-readable name (e.g., "B.Tech Computer Science & Engineering")
    url: str  # Fee structure URL
    timestamp: int = Field(default_factory=lambda: int(datetime.now().timestamp() * 1000))


class PlacementMessage(BaseModel):
    """Placement gallery trigger - opens full-screen placement photo viewer in Android app."""
    type: MessageType = MessageType.SHOW_PLACEMENTS
    title: str  # Title for the placement gallery (e.g., "Top Placements - GEHU Bhimtal")
    timestamp: int = Field(default_factory=lambda: int(datetime.now().timestamp() * 1000))


# ============== Conversation Memory ==============

class ConversationTurn(BaseModel):
    """A single turn in conversation."""
    role: str  # "user" or "assistant"
    content: str
    language: Language = Language.ENGLISH
    timestamp: datetime = Field(default_factory=datetime.now)


class ConversationHistory(BaseModel):
    """Conversation history for a session."""
    turns: List[ConversationTurn] = Field(default_factory=list)
    summary: Optional[str] = None
    total_tokens: int = 0


# ============== Message Parsing ==============

def parse_client_message(data: Dict[str, Any]) -> Optional[BaseModel]:
    """Parse incoming client message based on type."""
    msg_type = data.get("type")
    
    parsers = {
        MessageType.SESSION_START: SessionStartMessage,
        MessageType.SESSION_END: SessionEndMessage,
        MessageType.AUDIO_FRAME: AudioFrameMessage,
        MessageType.SPEECH_FINISHED: SpeechFinishedMessage,
        MessageType.INTERRUPT: InterruptMessage,
        MessageType.HEARTBEAT: HeartbeatMessage,
        MessageType.LANGUAGE_CHANGE: LanguageChangeMessage,
        MessageType.VOICE_CHANGE: VoiceChangeMessage,
        MessageType.TTS_PROVIDER_CHANGE: TtsProviderChangeMessage,
        MessageType.TTS_SPEED_CHANGE: TtsSpeedChangeMessage,
    }
    
    parser = parsers.get(msg_type)
    if parser:
        return parser(**data)
    return None
